subprojects/build-cache/src/main/java/org/gradle/caching/internal/controller/NextGenBuildCacheAccess.java

```diff
@@ -18,21 +18,169 @@
 
 import org.gradle.caching.BuildCacheKey;
 import org.gradle.caching.internal.NextGenBuildCacheService;
+import org.gradle.caching.internal.operations.BuildCacheRemoteLoadBuildOperationType;
+import org.gradle.caching.internal.operations.BuildCacheRemoteStoreBuildOperationType;
 
 import java.io.Closeable;
 import java.io.InputStream;
 import java.util.Map;
 
+/**
+ * Coordinates loading and storing cache entries in batches.
+ */
 public interface NextGenBuildCacheAccess extends Closeable {
+    /**
+     * Load the given entries if available.
+     *
+     * @param entries map of cache keys to load, associated with an optional payload to use during loading.
+     * @param handler the handler to call for entries loaded from the cache; passes the corresponding payload, too.
+     */
     <T> void load(Map<BuildCacheKey, T> entries, LoadHandler<T> handler);
 
+    /**
+     * Store the given entries in the build cache.
+     *
+     * @param entries map of cache keys to store, associated with an optional payload to use during storing.
+     * @param handler the handler to produce the bytes to store for each entry; passes the corresponding payload, too.
+     */
     <T> void store(Map<BuildCacheKey, T> entries, StoreHandler<T> handler);
 
     interface LoadHandler<T> {
+        /**
+         * Handle the loaded content.
+         *
+         * @param input the content loaded from the cache.
+         * @param payload optional payload belonging to this entry (see {@link #load(Map, LoadHandler)}).
+         */
         void handle(InputStream input, T payload);
+
+        /**
+         * Starts the legacy {@link BuildCacheRemoteLoadBuildOperationType} build operation when the first download starts.
+         */
+        void startLoadOperation(BuildCacheKey key);
```

Comment:
🤔 These method should probably belong to another interface, `BuildOpRecorder` or smth like that. 

Then we could pass it to load/store methods, e.g:
```
NextGenBuildCacheAccess.load(Map<BuildCacheKey, T> entries, BuildOpRecorder recorder, LoadHandler<T> handler);
```

And  since `BuildOpRecorder` would be passed as an argument to methods, we might simplify/remove `OperationFiringLoadHandlerFactory` and maybe also simplify synchronizations that we have now there. Maybe we can do something like that when we will have time to do some cleanup.

Reply:
FTR we agreed to do this when we go into production!